### PR TITLE
lib: add base16 (hex) encoding

### PR DESCRIPTION
**Pull request by simonvonhackewitz**
base16 (hexadecimal) as defined in [RFC 4648](https://datatracker.ietf.org/doc/html/rfc4648#section-8)

### CODE REVIEW

File: lib/encodings/base16.fz [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  # decode a valid base16 character to 4 bits
+  four_bits(n u8) =>
+    if n >= 48 && n <= 57        # case 0-9
+      n.as_u8 - 48
+    else #if n >= 65 && n <= 70  # case A-F
+      n.as_u8 - 55
+
+  # checks if a character is valid in the encoding
+  is_valid(c u8) =>
+    (c >= 48 && c <= 57) || (c >= 65 && c <= 70)
```
Comment (michaellilltokiwa) [inline lib/encodings/base16.fz:41]: does this work? `48 <= c <= 57`

Comment (simonvonhackewitz) [reply lib/encodings/base16.fz:41]: yes, nice

File: lib/encodings/base16.fz [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+public base16 is
+
+  alphabet array u8:= [48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 65, 66, 67, 68, 69, 70] #0123456789ABCDEF
+  encoding_name => "base16"
+
+  # decode a valid base16 character to 4 bits
+  four_bits(n u8) =>
+    if 48 <= n <= 57        # case 0-9
+      n.as_u8 - 48
+    else #if 65 <= n <= 70  # case A-F
```
Comment (michaellilltokiwa) [inline lib/encodings/base16.fz:36]: `else #if` seems weird?

File: lib/encodings/base16.fz [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+      res list u8 := nil, res ++ byte  # contains the decoded data at the end
+      nxt := 0, nxt + 2
+    while nxt < data.length
+    do
+      bits_0_3 := dec_char_at nxt
+      bits_4_7 := dec_char_at nxt+1
+
+      byte := if bits_0_3.ok && bits_4_7.ok
+                [(bits_0_3.val << 4) | bits_4_7.val]
+              else [(u8 0)]            # decoding result not used in error case
```
Comment (michaellilltokiwa) [inline lib/encodings/base16.fz:98]: [(u8 0)] =>  [u8 0]

File: lib/encodings/base16.fz [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+      n.as_u8 - 55
+
+  # checks if a character is valid in the encoding
+  is_valid(c u8) =>
+    (48 <= c <= 57) || (65 <= c <= 70)
+
+  # Encodes a given byte sequence in base16
+  # returns a sequence of ascii values
+  public encode(data array u8) =>
+    data.flat_map u8 (b->[alphabet[(b>>(u8 4)).as_i32], alphabet[(b&(u8 15)).as_i32]])
```
Comment (michaellilltokiwa) [inline lib/encodings/base16.fz:46]: b>>(u8 4) => b>>4

Comment (michaellilltokiwa) [reply lib/encodings/base16.fz:46]: b&(u8 15) => b&15

Comment (simonvonhackewitz) [reply lib/encodings/base16.fz:46]: Hm here it does actually work, but I've had the case where b was of type u64 and there it didn't work.

Comment (simonvonhackewitz) [reply lib/encodings/base16.fz:46]: Never mind there I've used a field instead of a constant

**Review by michaellilltokiwa — APPROVED** [review]

**Pull request merged**